test/e2e/syncer/spec_test.go

```diff
@@ -0,0 +1,254 @@
+				sourceCowboyClient := servers[sourceClusterName].wildwestClientSet.WildwestV1alpha1().Cowboys(testNamespace)
+				sinkNsClient := servers[sinkClusterName].k8sClientSet.CoreV1().Namespaces()
+				sinkWildwestClient := servers[sinkClusterName].wildwestClientSet.WildwestV1alpha1()
+
+				coreSharedInformerFactory := coreexternalversions.NewSharedInformerFactoryWithOptions(servers[sinkClusterName].k8sClientSet, 0)
+				nsInformer := coreSharedInformerFactory.Core().V1().Namespaces()
+				nsExpecter := framework.NewExpecter(nsInformer.Informer())
+				sinkNsLister := nsInformer.Lister()
+				coreSharedInformerFactory.Start(ctx.Done())
+
+				wildwestSharedInformerFactory := wildwestexternalversions.NewSharedInformerFactoryWithOptions(servers[sinkClusterName].wildwestClientSet, 0)
+				cowboyInformer := wildwestSharedInformerFactory.Wildwest().V1alpha1().Cowboys()
+				cowboyExpecter := framework.NewExpecter(cowboyInformer.Informer())
+				sinkCowboyLister := cowboyInformer.Lister().Cowboys("")
+				wildwestSharedInformerFactory.Start(ctx.Done())
+
+				ns, err := sourceNsClient.Create(ctx, &corev1.Namespace{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: testNamespace,
+					},
+				}, metav1.CreateOptions{})
+				if err != nil {
+					t.Errorf("failed to create namespace: %v", err)
+					return
+				}
+
+				cowboy, err := sourceCowboyClient.Create(ctx, &wildwestv1alpha1.Cowboy{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "timothy",
+					},
+					Spec: wildwestv1alpha1.CowboySpec{Intent: "yeehaw"},
+				}, metav1.CreateOptions{})
+				if err != nil {
+					t.Errorf("failed to create cowboy: %v", err)
+					return
+				}
+
+				if !cache.WaitForNamedCacheSync(t.Name(), ctx.Done(), nsInformer.Informer().HasSynced) {
+					t.Error("failed to wait for namespace cache to sync")
+					return
+				}
+
+				if !cache.WaitForNamedCacheSync(t.Name(), ctx.Done(), cowboyInformer.Informer().HasSynced) {
+					t.Error("failed to wait for cowboy cache to sync")
+					return
+				}
+
+				defer servers[sourceClusterName].Artifact(t, func() (runtime.Object, error) {
+					return sourceNsClient.Get(ctx, ns.Name, metav1.GetOptions{})
+				})
+				defer servers[sourceClusterName].Artifact(t, func() (runtime.Object, error) {
+					return sourceCowboyClient.Get(ctx, cowboy.Name, metav1.GetOptions{})
+				})
+				defer servers[sinkClusterName].Artifact(t, func() (runtime.Object, error) {
+					return sinkNsClient.Get(ctx, fmt.Sprintf("kcp--admin--%s", ns.Name), metav1.GetOptions{})
+				})
+				defer servers[sinkClusterName].Artifact(t, func() (runtime.Object, error) {
+					return sinkWildwestClient.Cowboys(fmt.Sprintf("kcp--admin--%s", cowboy.Name)).Get(ctx, cowboy.GetName(), metav1.GetOptions{})
+				})
+
+				if err := nsExpecter.ExpectBefore(ctx, func(ctx context.Context) (bool, error) {
+					sinkNsList, err := sinkNsLister.List(labels.Everything())
+					if err != nil {
+						return !apierrors.IsNotFound(err), err
```

Comment:
I don't believe you can get a NotFound error with a List call. I think you can safely return `true, err` because it's really hard (close to impossible) to get an error from a lister's List() call, and if we do get one, I doubt repeated attempts will yield different results.

Reply:
Most of them are doing a Get, not a List

Reply:
You are correct, my statement of 'all invocations' was in error. Still, the pattern appears to be metastasizing outside of the expect framework:

Get:
 - https://github.com/kcp-dev/kcp/blob/main/test/e2e/framework/expect.go#L238
 - https://github.com/kcp-dev/kcp/blob/main/test/e2e/framework/expect.go#L290

List:
 - https://github.com/kcp-dev/kcp/blob/main/test/e2e/reconciler/cluster/controller_test.go#L255
 - https://github.com/kcp-dev/kcp/blob/main/test/e2e/reconciler/ingress/controller_test.go#L343
 - https://github.com/kcp-dev/kcp/blob/main/test/e2e/framework/expect.go#L259
 - + the one in this PR